src/foam/dao/PostgresDAO.java

```diff
@@ -0,0 +1,175 @@
+
+        PgConnectionPool.setup(host, port, dbName, username, password);
+    }
+
+    @Override
+    public FObject remove(FObject o) {
+
+        try {
+
+            SQLData data = new SQLData(o);
+            Connection c = PgConnectionPool.getConnection();
+
+            PreparedStatement smt = c.prepareStatement(data.createDeleteStatement());
+
+            smt.setLong(1, ((Long) o.getProperty("id")));
+
+            int removed = smt.executeUpdate();
+            if (removed == 0) {
+                throw new SQLException("Error while removing.");
+            }
+            return o;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return null;
+    }
+
+    /**
+     * maps a database result row to an FObject
+     *
+     * @param row
+     * @return FObject
+     * @throws SQLException
+     */
+    private FObject createFObject(ResultSet row) throws Exception {
+
+        if (getOf() == null) {
+            throw new Exception("`Of` is not set");
+        }
+
+        List<PropertyInfo> props = getOf().getAxioms();
+        FObject result = (FObject) getOf().getObjClass().newInstance();
+
+        // set fields
+        for(int i = 0; i < props.size(); i++ ) {
+            String propName = props.get(i).getName();
+            Object value = row.getObject(i + 1);
+            result.setProperty(propName, value);
+        }
+        return result;
+    }
+
+    @Override
+    public FObject find(Object o) {
+
+        try {
+
+            String tableName = getOf().getObjClass().getSimpleName().toLowerCase();
+            Connection c = PgConnectionPool.getConnection();
+
+            String sql = "select * from " + tableName;
+            sql += " where id = ?";
+
+            PreparedStatement smt = c.prepareStatement(sql);
+            smt.setLong(1, ((Long) o));
+            ResultSet rs = smt.executeQuery();
+            if (rs.next()) {
+                return createFObject(rs);
+            }
+            return null;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    @Override
+    public FObject put(FObject obj) {
+
+        try {
+            Connection c = PgConnectionPool.getConnection();
+
+            // updating existing one
+            // when FObject isIdSet is null it returns 0 , is it ok ?
+            if ((Long) obj.getProperty("id") != 0) {
+                SQLData data = new SQLData(obj);
+                PreparedStatement smt = c.prepareStatement(data.createUpdateStatement());
+
+                // put data into statement preserving order.
+                int i = 1;
+                for (Object value: data.getValues()) {
+                    smt.setObject(i, value);
+                    i++;
+                }
+
+                int inserted = smt.executeUpdate();
+                if (inserted == 0) {
+                    throw new SQLException("Error while inserting.");
```

Comment:
If you throw this exception, or others, how is the connection returned to the pool?